### ISSUE TITLE
Encode uncompressed array to jpeg

### ISSUE THREAD

**Issue by kwmhp** [t0rakka/mango#71]
Hi,

I am trying to use the library to compress uncompressed color arrays to jpeg and png. Ideally I get another byte array out, so I can send the image to another service, but at first I've tried to save the bitmap to a file. 
The code is rather simple and my jpg can be opened and has the correct dimensions, but the image is solid black.

My Image is an vector of a Color struct.

```
Color {
uint8 B;
uint8 G;
uint8 R;
uint8 A;
}


const mango::Surface surface(
	Width, Height, mango::Format(32, mango::Format::UINT, mango::Format::BGRA, 8, 8, 8, 8), 4 * Width,  Image.data());
mango::Bitmap bitmap(surface, mango::Format(32, mango::Format::UINT, mango::Format::BGRA, 8, 8, 8, 8));

mango::ImageEncodeOptions options;
options.quality = 0.80;

bitmap.save("output-mango.jpg", options);
```

I guess the stride or format is incorrect.  But I could figure it out so far.

Thanks for any kind of help, and especially thank you for providing this amazing library 👍 

Kim

**Comment by t0rakka** [comment]
The Format is incorrect, UINT means the values are not normalized: 8 means literally 8.0. You want to use UNORM, which means "unsigned normalized" so if maximum value is 255, then literal value of 255 will map into 1.0 and so on.

You don't have to make a copy of your input data; you can pass the Surface to encoder. The naming of these classes are a bit ancient, here's what they mean:

Bitmap is a container, which "owns" the memory.

Surface is kind of a "BitmapView", view into a bitmap. It simply describes the dimensions of image rectangle, stride, pixel format so that you can access raw memory as-if it were an image.

In modern APIs "Surface" owns the memory and "SurfaceView" is just a view into a surface, but it's a bit late to change now.. OK, now that we know what's what, when you create the bitmap object in your code it will actually allocate storage and blit from the surface into the bitmap. You don't need to do this for the encoder; it will do any color conversion it needs itself.. just pass it a surface and it can do its trick. If the surface is in format already supported by the encoder it will skip the color conversion.

32 bit RGBA and BGRA are formats that most encoders can encode directly from. BGRA is probably more commonly supported but don't worry about it too much, it's just efficiency issue.. other formats will mostly work, too. FLOAT and UNORM are best supported, for example, in your case you pass UINT data there isn't even valid conversion to UNORM so the blitter can't do anything. UNORM 8 bit channels have values between 0.0 and 255.0, where the JPEG/PNG can only encode from 0.0 to 1.0 (1.0 being the maximum brightness, obviously).

TLDR;

1. use Format::UNORM
2. encode directly from the surface (which just tells where and what your data is)

If you look at include/image/encoder.hpp, you notice that the ImageEncoder target is "Stream&", so it can be a file, or in your case you might want s MemoryStream, which is in core/buffer.hpp

It has following constructors:

        MemoryStream();
        MemoryStream(const u8* source, size_t bytes);
        MemoryStream(ConstMemory memory);
        ~MemoryStream();

In your case you want to use the default constructor so that in the beginning the stream is empty, then the encoder will add bytes into it, which you can do what you want.. send over network, write into a file, whatever.

It might make sense to add convenience member functions into Surface to be able to do: surface.save(stream, options) since it is fairly common thing to do.. but for now you have to use the ImageEncoder class to get the job done, good luck! :)

**Comment by kwmhp** [comment]
Wow, thanks for the quick and detailed reply.

Changing the format to  Format::UNORM did the trick. I've looked over the formats but didn't draw the connection to normalized.

Encoding the surface to the MemoryStream worked flawlessly. I had some of the code already setup but I thought I had to allocate a buffer myself. 
I've replaced it with the default ctor and now I have a working encoding within 6 lines of code. Perfect!

**Comment by t0rakka** [comment]
Glad it wasn't some blitter bug or anything of that sort. I'm slowly upgrading the blitter, did some foundation work already if you look at the commits. :P 

The biggest roadblock is that vector shifts where each lane can have a different shift count are AVX2 only, emulating that with earlier ISA's is 10+ instructions, which means slower code, which means I am still not happy with the solution I have. For AVX2 it's really cool since the new color conversion engine is super precise (and fast when have SIMD). It's based on integer approach for doing  float(src_bits) / float(dest_bits), there is bit-precise way to do it in integer.. but requires per lane variable shifting... :_(  I have a trick in mind to work around that, which is unrolling component handling, so I do N red channels, N green channels, and so on, then finally combine the result.

But.. before it's tested and profiled, got nothing..

**Comment by kwmhp** [comment]
This goes way over my head but I love the passion you have for optimized code. 

I'm working on a latency sensitive application and switching from Matt Geldreich's jpeg-compressor to mango saved me up to 20% from user input to a jpg. And he is a image compression guru by himself.
The only downside seems to be the image size. With the jpeg-compressor's default quality of 0.85, mango images are up to 70% larger. 

I will have to try different quality settings for both encoders and compare the images with comparable filesize.
